sympy/functions/combinatorial/factorials.py

```diff
@@ -240,32 +271,43 @@ def _eval_rewrite_as_gamma(self, n, **kwargs):
         return gamma(n + 1)
 
     def _eval_rewrite_as_Product(self, n, **kwargs):
-        from sympy import Product
-        if n.is_nonnegative and n.is_integer:
-            i = Dummy('i', integer=True)
+        from sympy.concrete.products import Product
+        if n.is_nonnegative and int_like(n):
+            for i in igen():
+                if not n.has(i):
+                    break
```

Comment:
Why isn't this just using Dummy?

Reply:
Internally they are fine, but they look bad for public consumption.